File: cli/src/core/references/sources/GitHubNormalize.ts
/**
 * GitHubNormalize — GitHub-domain payload reshaping, owned by the GitHub source
 * and shared by every producer that yields GitHub issues in a non-canonical
 * shape (the Codex `codex_apps` connector today; the `gh` CLI later).
 *
 * Maps a producer's issue object into the shape `GitHubAdapter.extractRef` reads:
 * unwrap `issue.*`, `issue_number`→`number`, `url`→`html_url`, flatten the
 * object-array `labels`/`assignees` into string arrays, and — for search hits
 * that leave `number` null — derive the issue number from the URL.
 *
 * Self-contained (local `isObject`, no cross-layer import) so `sources/` never
 * depends on `bindings/`.
 */

const ISSUE_NUMBER_IN_URL = /\/(?:issues|pull)\/(\d+)/;

function isObject(v: unknown): v is Record<string, unknown> {
	return typeof v === "object" && v !== null && !Array.isArray(v);
}

/** Flatten an array of `{[key]: string}` objects (or bare strings) to a string array. */
function flattenNamed(value: unknown, key: "name" | "login"): string[] | undefined {
	if (!Array.isArray(value)) return undefined;
	const out: string[] = [];
	for (const item of value) {
		if (typeof item === "string" && item.length > 0) {
			out.push(item);
		} else if (isObject(item)) {
			const v = item[key];
			if (typeof v === "string" && v.length > 0) out.push(v);
		}
	}
	return out.length > 0 ? out : undefined;
}

/** Reshape one GitHub issue (single fetch OR one search-result element) into adapter shape. */
export function reshapeGitHubIssue(raw: unknown): unknown {
	if (!isObject(raw)) return raw;
	const issue = isObject(raw.issue) ? raw.issue : raw;
	const out: Record<string, unknown> = {};

	const num = issue.issue_number ?? issue.number;
	if (typeof num === "number") out.number = num;
	if (typeof issue.title === "string") out.title = issue.title;
	const url = issue.url ?? issue.html_url;
	if (typeof url === "string") out.html_url = url;
	if (typeof issue.body === "string") out.body = issue.body;
	if (typeof issue.state === "string") out.state = issue.state;

	const labels = flattenNamed(issue.labels, "name");
	if (labels !== undefined) out.labels = labels;
	const assignees = flattenNamed(issue.assignees, "login");
	if (assignees !== undefined) out.assignees = assignees;

	const fullName = issue.repository_full_name ?? raw.repository_full_name;
	if (typeof fullName === "string") out.repository = { full_name: fullName };

	// Search hits leave `number` null but always carry the issue URL — derive the
	// issue number from it so the adapter (which keys nativeId off `number`) accepts it.
	if (out.number === undefined && typeof out.html_url === "string") {
		const m = ISSUE_NUMBER_IN_URL.exec(out.html_url);
		if (m) out.number = Number(m[1]);
	}

	return out;
}

File: intellij/src/main/kotlin/ai/jolli/jollimemory/core/references/GitHubNormalize.kt
package ai.jolli.jollimemory.core.references

import com.google.gson.JsonArray
import com.google.gson.JsonElement
import com.google.gson.JsonObject

/**
 * GitHub-domain payload reshaping.
 *
 * Maps a producer's issue object into the shape [GitHubAdapter.extractRef] reads:
 * unwrap `issue.*`, `issue_number`→`number`, `url`→`html_url`, flatten the
 * object-array `labels`/`assignees` into string arrays, and — for search hits
 * that leave `number` null — derive the issue number from the URL.
 */
object GitHubNormalize {

	private val ISSUE_NUMBER_IN_URL = Regex("/(?:issues|pull)/(\\d+)")

	/** Flatten an array of `{[key]: string}` objects (or bare strings) to a string JsonArray. */
	private fun flattenNamed(value: JsonElement?, key: String): JsonArray? {
		if (value == null || !value.isJsonArray) return null
		val arr = value.asJsonArray
		val out = JsonArray()
		for (item in arr) {
			if (item.isJsonPrimitive && item.asJsonPrimitive.isString) {
				val s = item.asString
				if (s.isNotEmpty()) out.add(s)
			} else if (item.isJsonObject) {
				val v = item.asJsonObject.stringOrNull(key)
				if (!v.isNullOrEmpty()) out.add(v)
			}
		}
		return if (out.size() > 0) out else null
	}

	/** Reshape one GitHub issue (single fetch OR one search-result element) into adapter shape. */
	fun reshape(raw: JsonElement?): JsonElement? {
		if (raw == null || !raw.isJsonObject) return raw
		val rawObj = raw.asJsonObject
		val issue = rawObj.objectOrNull("issue") ?: rawObj
		val out = JsonObject()

		val num = issue.intOrNull("issue_number") ?: issue.intOrNull("number")
		if (num != null) out.addProperty("number", num)
		issue.stringOrNull("title")?.let { out.addProperty("title", it) }
		val url = issue.stringOrNull("url") ?: issue.stringOrNull("html_url")
		if (url != null) out.addProperty("html_url", url)
		issue.stringOrNull("body")?.let { out.addProperty("body", it) }
		issue.stringOrNull("state")?.let { out.addProperty("state", it) }

		flattenNamed(issue.get("labels"), "name")?.let { out.add("labels", it) }
		flattenNamed(issue.get("assignees"), "login")?.let { out.add("assignees", it) }

		val fullName = issue.stringOrNull("repository_full_name")
			?: rawObj.stringOrNull("repository_full_name")
		if (fullName != null) {
			val repo = JsonObject()
			repo.addProperty("full_name", fullName)
			out.add("repository", repo)
		}

		// Search hits leave `number` null — derive from URL.
		if (!out.has("number") || !out.get("number").isJsonPrimitive) {
			val htmlUrl = out.stringOrNull("html_url")
			if (htmlUrl != null) {
				val m = ISSUE_NUMBER_IN_URL.find(htmlUrl)
				if (m != null) out.addProperty("number", m.groupValues[1].toInt())
			}
		}

		return out
	}
}
